src/collection-view.js

```diff
@@ -645,18 +654,25 @@ const CollectionView = Backbone.View.extend({
 
     const hasIndex = (typeof index !== 'undefined');
 
-    // Only cache views if added to the end
-    if (!hasIndex || index >= this._children.length) {
+    // Only cache views if added to the end and there is no awaiting views
+    if (!preventRender && !this._addingMultipleViews && (!hasIndex || index >= this._children.length)) {
```

Comment:
this still will not work because renders can occur separate from calling `addChildView`  any render that occurs apart from `addChildView` if `this._addedViews` exists will be broken.

Reply:
i think i not follow
_addedViews will be used only if this is call with `preventRender: false` and there is no awaiting views. in other cases _addedViews will not be used at all
or i do not see something?

Reply:
ah I see.  yes I think you're mostly correct.

There is an unexpected side-effect though.  If `preventRender` was called and `_addingMultipleViews` is set, and then the collection adds views to the end it will trigger the `_addedViews` perf and only render those views without the unrendered view, which is not good.  I think maybe this flag should change to `_hasUnrenderedViews`.  Additionally the flag needs to be unset if the collection is re-rendered by some other way as well.

So I think perhaps the first line of `_renderChildren` should be:
```
if (this._hasUnrenderedViews) {
  delete this._addedViews;
  delete this._hasUnrenderedViews;
}
```

Had an idea though. on collection update `_addChild` happens first, which makes sense here as well.
This gets simpler if it is:
- `_addChild`
- `if preventRender return`
- `_addedViews =`

Reply:
did you mean this case?
```
const test = new ExtendedAddChildView({
	collection,
	childView: ModelView,
	viewComparator: false,
	onRender(){
		let preventRender = true;
		this.addChildView(new CustomView({ text:'c1' }), { preventRender });
    this.collection.add({id:4});
		this.addChildView(new CustomView({ text:'c2' }), { preventRender });
		this.addChildView(new CustomView({ text:'c3' }), 3, { preventRender });
		this.addChildView(new CustomView({ text:'c4' }), undefined, { preventRender });
		this.addChildView(new CustomView({ text:'c5' }));
	},
	
});
```
if so, then all views will be added
i think i am not understand the case
here is a codepen for this pull request: https://codepen.io/dimatabu/pen/jvMmxp

Reply:
Results show `2, 1, 4`
```js
	onRender(){
		this.addChildView(new CustomView({ text:'c1' }), { preventRender: true });
    this.collection.set([{id:1},{id:2},{id:4}]);
}
```
Results show `c1, 2, 1`
```js
	onRender(){
		this.addChildView(new CustomView({ text:'c1' }), { preventRender: true });
    this.collection.set([{id:1},{id:2}]);
}
```

Reply:
that's actually a _different_ issue than earlier mentioned.. still looking for other potential issues

Reply:
yes, but there should call with preventRender: false
i mean, if you do:
addChildView (prevent:true)
addChildView (prevent:true)
... here something happens, say collection receive new models, and renderchildrens occurs
addChildView(prevent:true)
addChildView()
done. here renderChildren occurs and all views in own places
am i right?

Reply:
i mean, if you add a view with preventRender it means that you must add another view, the final one without `preventRender: true` for putting views in a place

Reply:
but maybe you want to tell that collection update event should be treaten as finalize in adding multiple views?

Reply:
I'm completely against adding this if it requires that you call the function a particular way at the end.  That makes no sense to me.
What should happen is that no matter what else causes a render the behavior should be the same.
Consider:  (also note the order of things)
```js
const ExtendedAddChildView = Mn.CollectionView.extend({
  addChildView(view, index, options = {}) {
    if (!view || view._isDestroyed) {
      return view;
    }
    if (_.isObject(index)) {
      options = index;
    }
    // If options has defined index we should use it
    if (options.index != null) {
      index = options.index;
    }

    this._addChild(view, index);

    // `preventRender` passed, so we are waiting another `addChildView`
    if (options.preventRender) {
      this._hasUnrenderedViews = true;
      return view;
    }
    
    if (!this._isRendered) {
      this.render();
    }
    
    const hasIndex = (typeof index !== 'undefined');
    const isAddedToEnd = !hasIndex || index >= this._children.length;

    // Only cache views if added to the end and there is no unrendered views
    if (isAddedToEnd && !this._hasUnrenderedViews) {
      this._addedViews = [view];
    }
    
    if (hasIndex) {
      this._renderChildren();
    } else {
      this.sort();
    }
    
    return view;
  },
  _renderChildren() {
    // If there are unrendered views prevent add to end perf
    if (this._hasUnrenderedViews) {
      delete this._addedViews;
      delete this._hasUnrenderedViews;
    }
    Mn.CollectionView.prototype._renderChildren.call(this);
  }
});
```

Reply:
i've tried that
but now two test fails, trying to resolve this
*when the collectionView is not attached*
```
      it('should trigger "before:render" event on the childView', function() {
        expect(childView.onBeforeRender).to.have.been.calledOnce.and.calledWith(childView);
      });

      it('should trigger "render" event on the childView', function() {
        expect(childView.onRender).to.have.been.calledOnce.and.calledWith(childView);
      });
```

Reply:
```
    if (!this._isRendered) {
      this.render();
    }
    this._addChild(view, index);

    // `preventRender` passed, so we are waiting another `addChildView`
    if (options.preventRender) {
      this._hasUnrenderedViews = true;
      return view;
    }
```
otherwise we get two filed test in render child lifecocle mentioned above

Reply:
ah yep.  that makes sense